openedx/core/djangoapps/credentials/tasks/v1/tasks.py

```diff
@@ -54,3 +75,288 @@ def send_grade_to_credentials(self, username, course_run_key, verified, letter_g
+            'course_key': cert.course_id,
+            'mode': cert.mode,
+            'status': cert.status,
+            'verbose': verbose,
+        }
+
+        data = {
+            'mode': cert.mode,
+            'status': cert.status
+        }
+
+        course_cert_info[(cert.user.id, str(cert.course_id))] = data
+        handle_course_cert_changed(**signal_args)
+        if notify_programs and CertificateStatuses.is_passing_status(cert.status):
+            handle_course_cert_awarded(**signal_args)
+
+    # Then do grades
+    for i, grade in paged_query(grades, delay, page_size):
+        if site_config and not site_config.has_org(grade.course_id.org):
+            logger.info("Skipping grade changes %d for grade %s", i, gradestr(grade))
+            continue
+
+        logger.info(
+            "Handling grade changes %d for grade %s",
+            i, gradestr(grade),
+        )
+
+        user = User.objects.get(id=grade.user_id)
+
+        # Grab mode/status from cert call
+        key = (user.id, str(grade.course_id))
+        cert_info = course_cert_info.get(key, {})
+        mode = cert_info.get('mode', None)
+        status = cert_info.get('status', None)
+
+        send_grade_if_interesting(
+            user,
+            grade.course_id,
+            mode,
+            status,
+            grade.letter_grade,
+            grade.percent_grade,
+            verbose=verbose
+        )
+
+
+def paged_query(queryset, delay, page_size):
+    """
+    A generator that iterates through a queryset but only resolves chunks of it at once, to avoid overwhelming memory
+    with a giant query. Also adds an optional delay between yields, to help with load.
+    """
+    count = queryset.count()
+    pages = int(math.ceil(count / page_size))
+
+    for page in range(pages):
+        page_start = page * page_size
+        page_end = page_start + page_size
+        subquery = queryset[page_start:page_end]
+
+        if delay and page:
+            time.sleep(delay)
+        index = 0
+        try:
+            for item in subquery.iterator():
+                index += 1
+                yield page_start + index, item
+        except OperationalError:
+            # When running the notify_credentials command locally there is an
+            # OperationalError thrown by MySQL when there are no more results
+            # available for the queryset iterator. This change catches that exception,
+            # checks state, and then logs according to that state. This code runs in
+            # production without issue. This changes allows for the code to be run
+            # locally without a separate code path.
+            if index == count:
+                logger.info('OperationalError Exception caught, all known results processed in paged_query')
+            else:
+                logger.warning('OperationalError Exception caught, it is possible some results were missed')
+            continue
+
+
+def log_dry_run(certs, grades):
```

Comment:
This method changed from `print_dry_run` to `log_dry_run`, and what were previously `print` statements have been changed to `logger.info` statements so we can at least see them somewhere. This in response to Matt pointing out:

> Since this is all happening in a task, it won't show up in the Jenkins console (or print to local console). This really minimizes the benefit of a dry run option, but we don't need to refactor now. Instead can we just make this function log instead of print and change "command" to "task"? ([Source](https://github.com/edx/edx-platform/pull/27322#discussion_r612702566))